### DISCUSSION TITLE
Enable Cargo feature in BUILD.bazel rust_test

### DISCUSSION THREAD

**Discussion by antjori** [bazelbuild/rules_rust#1869] [Q&A] [answered]
Hey everyone!

I have a `Cargo.toml` file where I've defined a Cargo feature:

```
[features]
my-feature: []
```

I'm trying to enable this feature but only for an integration test without much success. I was expecting the following to work:

```
rust_library(
    name = "my-lib",
    srcs = [
        (...)
    ],
    deps = [
         (...)
    ],
)

rust_test(
    name = "integration-test",
    srcs = glob(["tests/**"]),
    crate_features = [
        "my-feature",
    ],
    deps = [
        ":my-lib",
    ],
)
```

If I add the `crate_features` attribute to the `rust_library` though, it works. Unfortunately, by doing it, the feature becomes available to the entire binary which is not intended.

What am I missing?

**Comment by scentini** [comment]
What is the error you're seeing when building what you spelled out above?

**Reply by antjori** [reply]
No error whatsoever. Only the Cargo feature isn't applied when the Bazel tests run. Whereas, if I move the attribute `crate_features` to the `rust_library` def and run the Bazel tests, the feature is taken into account.

**Comment by illicitonion** [comment, marked as answer]
This is working as intended - `crate_feature` applies to the specific code it's compiling and nothing else.

If you want to achieve what you're trying to do, you'd need a second copy of the `rust_library` which enables the feature, and to have the `rust_test` depend on that.

(Alternatively, you could potentially use a transition to alter the `rust_library`, but the effect is the same - effectively splitting the `rust_library` into two targets)

**Reply by antjori** [reply]
Thank you for your suggestion!

That was exactly what I tried to do but without any luck. That's why I opened this discussion.

Since the second copy of the `rust_library` cannot have the same name as the original one, and since the `rust_test` imports dependencies that reference the original `rust_library`, the `rustc` complains saying I should reference the second copy of the `rust_library`. But if I change the import dependencies from the original lib to the second copy, then the `rustc` complains as well because, for all that matters, the second copy's name isn't defined as being the lib.

I also tried creating the second copy of the `rust_library` with only the `name` and `crate_features` attributes, and then I'd add both libraries as `deps` of the `rust_test`, but Bazel complains saying that no source file was found in the second copy definition. I'm assuming the `srcs` attribute is required even though the documentation says it is `Optional`.

Is there a way to make both coexist given the details mentioned above?

**Reply by illicitonion** [reply]
You can use http://bazelbuild.github.io/rules_rust/flatten.html#rust_library-crate_name to change the name to be the same, but you're going to need to ensure somehow that a binary/test doesn't try to depend on both versions of the rust_library, or you'll end up with link-errors because of duplicate symbols.

**Reply by antjori** [reply]
Thank you so much, it worked! The test only depends on the secondary copy of the `rust_library` so no errors occurred. It was just that `crate_name` piece that was missing! 🙌🏼
